### ISSUE TITLE
Cookie doesn't seem to work 

### ISSUE THREAD

**Issue by karelrooted** [b5i/YouTubeKit#4]
I try add these three cookie YTM.cookies = "SAPISID=\(SAPISID); __Secure-1PAPISID=\(PAPISID); __Secure-1PSID=\(PSID1)"
or all the cookie in safari YouTube official website (host: MacBook)

also tried include some other  header currently not present in the code , did it work on your end as of now ?

**Comment by b5i** [comment]
Hello, thanks for raising this issue!
> I try add these three cookie YTM.cookies = "SAPISID=\(SAPISID); __Secure-1PAPISID=\(PAPISID); __Secure-1PSID=\(PSID1)"
Ok this should work, the procedure is currently done in Atwy [here](https://github.com/b5i/Atwy/blob/41d19179ce846ff2f8bcc6d4ba315c4cc1e2ab69/Atwy/WebView.swift#L111). Maybe check if their domain is `domain:.youtube.com` and not `google.com` for example. Let me know if that fixes your issue.
> 
> or all the cookie in safari YouTube official website (host: MacBook)
This could not work as all of them are mixed and YouTube might interpret them in a wrong way.
> 
> 
> 
> also tried include some other  header currently not present in the code , did it work on your end as of now ? 
Yes, you can check out [Atwy](https://github.com/b5i/Atwy), this is an App To Watch YouTube, and it uses all the features of YouTubeKit.

**Comment by karelrooted** [comment]
I think I found the problem , I log in YouTube with multiple account, when log in with multiple account , these three cookie (SAPISID, __Secure-1PAPISID,__Secure-1PSID) are the same with each account,  we need to add another three cookie to identify the current user  (  LOGIN_INFO, __Secure-1PSIDTS,  __Secure-3PSIDTS=\(PSIDTS1) ) 

in the short time I test, the later three cookie seems to change in a very short time, so it seems it does not work (when I test yesterday with all the cookie and did not work , it's highly likely the later three cookie change and that time)

**Comment by b5i** [comment]
> I think I found the problem , I log in YouTube with multiple account, when log in with multiple account , these three cookie (SAPISID, __Secure-1PAPISID,__Secure-1PSID) are the same with each account,  we need to add another three cookie to identify the current user  (  LOGIN_INFO, __Secure-1PSIDTS,  __Secure-3PSIDTS ) 

Very interesting, I'm also going to check later in the morning that case and try to find an solution.

**Comment by b5i** [comment]
> I use youtubekit to develop a player for personal use on Apple TV (to enable bilingual subtile support), I don't think there is YouTube cookie to fetch on Apple TV (webview is not allowed on Apple TV, will check it later )

As your project is designed for personal use you could use the private API as https://github.com/Moballo-LLC/tvOS-Browser does, its should cause any problem (I already tested on my Apple TV and it worked fine).

You could also do a kind of shared keychain (like Atwy is going to do for its TV app), like you have a little app on your iPhone that will fetch the cookies and put them in a shared environnement.

**Comment by karelrooted** [comment]
> > I use youtubekit to develop a player for personal use on Apple TV (to enable bilingual subtile support), I don't think there is YouTube cookie to fetch on Apple TV (webview is not allowed on Apple TV, will check it later )
> 
> As your project is designed for personal use you could use the private API as https://github.com/Moballo-LLC/tvOS-Browser does, its shouldn't cause any problem (I already tested on my Apple TV and it worked fine).
> 
> You could also do a kind of shared keychain (like Atwy is going to do for its TV app), like you have a little app on your iPhone that will fetch the cookies and put them in a shared environnement (or even just in your clipboard and you enter the cookies string in a prompt of your iPhone's remote app).

if I close the YouTube website in safari  , the cookie seems to work for a longer time, I will try your suggestion when cookie expire too often 

by the way , to choose current user other than default (in multiple account scenario), additional header need to be added, please check my updated comment above

**Comment by b5i** [comment]
What do you think about new methods to the `YouTubeModel` that lets you select an account with a bunch of cookies provided?

**Comment by b5i** [comment]
> I think I found the problem , I log in YouTube with multiple account, when log in with multiple account , these three cookie (SAPISID, __Secure-1PAPISID,__Secure-1PSID) are the same with each account, we need to add additional two cookie ( LOGIN_INFO, __Secure-1PSIDTS )

For this I just tested again and it seems to work fine, except the fact that it connects you with the first account that has been registered in the browser (and then there would be the procedure with `X-Goog-AuthUser` and all of that). Did it connect you to an account when you tested only with the 3 (SAPISID, __Secure-1PAPISID,__Secure-1PSID) cookies?

**Comment by karelrooted** [comment]
> What do you think about new methods of `YouTubeModel` that lets you select an account with a bunch of cookies provided?

sounds good,  in my current two account setup:

the first account is  1PSID,  x-goog-authuser =0
the second accounts 3PS1D,   x-goog-authuser =1

if the order is the same in cookie and  x-goog-authuser, then it should work without some additional work(auto switch account to get the account's header info ) ( I tried delete the x-goog-visitor-id header, the account is still currently selected )

**Comment by karelrooted** [comment]
> > I think I found the problem , I log in YouTube with multiple account, when log in with multiple account , these three cookie (SAPISID, __Secure-1PAPISID,__Secure-1PSID) are the same with each account, we need to add additional two cookie ( LOGIN_INFO, __Secure-1PSIDTS )
> 
> For this I just tested again and it seems to work fine, except the fact that it connects you with the first account that has been registered in the browser (and then there would be the procedure with `X-Goog-AuthUser` and all of that). Did it connect you to an account when you tested only with the 3 (SAPISID, __Secure-1PAPISID,__Secure-1PSID) cookies?

no, if only the first three cookie is provided, google response with please login (in user library request), in the two account I login (one is YouTube premium, may be something is different with premium account)

**Comment by b5i** [comment]
> no, if only the first three cookie is provided, google response with please login (in user library request), in the two account I login (one is YouTube premium, may be something is different with premium account)
> 
> * the 1PSID is the no premium account

Are __Secure-1PAPISID and __Secure-3PAPISID equals on your side?

**Comment by karelrooted** [comment]
> > no, if only the first three cookie is provided, google response with please login (in user library request), in the two account I login (one is YouTube premium, may be something is different with premium account)
> > 
> > * the 1PSID is the no premium account
> 
> Are __Secure-1PAPISID and __Secure-3PAPISID equals on your side?

yes, they are the same

by the way: after delete login_info cookie, it still works, so the additional cookie needed is  __Secure-{number}PSIDTS

**Comment by b5i** [comment]
> yes, they are the same
> 
> by the way: after delete login_info cookie, it still works, so the additional cookie needed is __Secure-{number}PSIDTS

This is weird because when compare the cookies before and after switching `__Secure-{number}PSIDTS` is still the same.

**Comment by karelrooted** [comment]
> > yes, they are the same
> > by the way: after delete login_info cookie, it still works, so the additional cookie needed is __Secure-{number}PSIDTS
> 
> This is weird because when compare the cookies before and after switching `__Secure-{number}PSIDTS` is still the same.

yes,  __Secure-1PSIDTS and  __Secure-3PSIDTS are the same too, it seems when one of the account is premium, then google need this additional cookie, regardless which account we are trying to access
